### DISCUSSION TITLE
SEO Crawl Analysis Template : question

### DISCUSSION THREAD

**Discussion by httpwebmedia** [eliasdabbas/advertools#283] [General]
Hello Elias, I wished I could contact you through mail but for some reasons I can't DM people on twitter.

So here I am and ask you a little help on your great tool and more specifically on https://www.kaggle.com/code/eliasdabbas/seo-crawl-analysis-template

I've been watching your video https://youtu.be/IZUBvDfDbqo

and I would like to make the code works with the following interactive tool : 
![image](https://github.com/eliasdabbas/advertools/assets/21228346/e3602f7f-2b47-4c0a-8b36-b5b475bd92fe)

In Jupyter notebook, when you execute the code : 
def count_column_values(col_name, show_top=10, sort_others=False):
    series = crawl_df[col_name].str.split('@@').explode()
    print(f'Number of {series.name}s: {series.count():,}')
    print(f'Number of pages: {len(crawl_df):,}')
    print(f'{series.name}s per page: {series.count()/len(crawl_df):.2f}')
    display(value_counts_plus(series, sort_others=sort_others, dropna=False, show_top=show_top))

widgets.interact(count_column_values,
                 col_name=crawl_df.select_dtypes('object').columns,
                 show_top=widgets.IntSlider(min=1, max=50, value=10));

it returns a line partially hidden like this : 
![image](https://github.com/eliasdabbas/advertools/assets/21228346/40546a24-d861-4fb5-bc3c-c86adb463ac1)

So my question is, how do I get the interactive code  like in your video please?

Thanks a lot for your great work!!

**Comment by httpwebmedia** [comment]
I forgot to mention, if I run the widget code in python interpreter, it returns this line : 
interactive(children=(Dropdown(description='col_name', options=('url', 'title', 'meta_desc', 'viewport', 'charset', 'h1', 'h2', 'h3', 'canonical', 'alt_href', 'og:site_name', 'og:type', 'og:description', 'og:title', 'og:url', 'og:image', 'twitter:site', 'twitter:image', 'twitter:card', 'jsonld_@context', 'jsonld_@graph', 'body_text', 'download_slot', 'links_url', 'links_text', 'links_nofollow', 'nav_links_url', 'nav_links_text', 'nav_links_nofollow', 'header_links_url', 'header_links_text', 'header_links_nofollow', 'footer_links_url', 'footer_links_text', 'footer_links_nofollow', 'img_loading', 'img_src', 'img_srcset', 'img_alt', 'img_width', 'img_sizes', 'img_height', 'ip_address', 'resp_headers_date', 'resp_headers_content-type', 'resp_headers_link', 'resp_headers_x-cache', 'resp_headers_x-server', 'resp_headers_cache-control', 'resp_headers_cf-cache-status', 'resp_headers_last-modified', 'resp_headers_report-to', 'resp_headers_nel', 'resp_headers_vary', 'resp_headers_server', 'resp_headers_cf-ray', 'request_headers_accept', 'request_headers_accept-language', 'request_headers_user-agent', 'request_headers_accept-encoding', 'redirect_urls', 'h4', 'h5'), value='url'), IntSlider(value=10, description='show_top', max=50, min=1), Checkbox(value=False, description='sort_others'), Output()), _dom_classes=('widget-interact',))

but I can't run it in Jupyter...

**Reply by eliasdabbas** [reply]
Thanks @httpwebmedia 

Can you please tell me a bit more:

1. What is the exact difference between the video and what you are getting?
2. Where are you running the code, on Kaggle, on your laptop, somewhere else?
3. What package versions are you using for advertools pandas scrapy?

**Reply by httpwebmedia** [reply]
Thanks for your response :) 

1. Like I said in my first post, I can't get the interactive dropdown and slider. ( have a look at my first screenshot I highlighted the dropdown and slider. This part is not displayed on Jypiter notebook, it returns the interactive line I wrote on my second post...
2.  I run the code on Jupyter notebook on my windows 11 pc.

jupyter --version
Selected Jupyter core packages...
IPython          : 8.13.2
ipykernel        : 6.23.1
ipywidgets       : 8.0.6
jupyter_client   : 7.4.1
jupyter_core     : 5.3.0
jupyter_server   : 1.24.0
jupyterlab       : 4.0.0
nbclient         : 0.7.4
nbconvert        : 7.4.0
nbformat         : 5.8.0
notebook         : 6.5.4
qtconsole        : not installed
traitlets        : 5.9.0

3. >>> print(pd.__version__)
1.4.2
>>> print(adv.__version__)
0.13.2

I don't use scrapy...

Hope this helps !

**Reply by eliasdabbas** [reply]
Sorry I thought that was already working on your machine. 

From what you are saying it seems that there is an issue with `ipywidgets` because it is the package responsible for the interactive widgets.

Here is the issue:

https://github.com/jupyter-widgets/ipywidgets/issues/3781

You seem to be using version 4.0.0. Maybe try to install  v3.6.3?

Please let me know how it goes.

**Reply by httpwebmedia** [reply]
I downgraded as per your advice and it worked :D 

thanks a lot!!!

![image](https://github.com/eliasdabbas/advertools/assets/21228346/6f7ea3c2-5d6a-43a3-ba43-8f8ed8517aa6)

Besides of it, I found an issue with advertool crawler. I used it on a website that uses many title html tags <title></title> within the same page. This results of a similar output  : 
```
crawl_df[:1][['title']]

                                                                                                                                                                     title
0  The world's number one website | Testsite.com@@testsite.com@@Search@@Facebook@@Twitter@@YouTube@@Reddit@@RSS@@Right Arrow@@Right Arrow
```

I thought the title column would be the title tag within the header tag. 

Any thought?

**Reply by eliasdabbas** [reply]
1. Great to know it worked! Enjoy the interactivity.
2. I think the crawler is behaving correctly. Otherwise you wouldn't have known about this issue. You can use custom extraction, XPath or CSS selectors, if you want only title tags within the head. Does that make sense?

**Reply by httpwebmedia** [reply]
yes, it makes sense :) thanks for all!

One last, will you make the chat gpt entity extractor code available?

**Reply by eliasdabbas** [reply]
Great! Happy it worked, and I'm still not sure why on earth the Economist has so many titles on their pages.

Yes, I've written an article about how to train ChatGPT for the entity extraction app and it will be published hopefully soon.

**Reply by httpwebmedia** [reply]
Hello Elias, I still have great fun using your great tool.

I was wondering how I could possibly use https proxies with robotstxt_to_df?
I'm blocked by the site if I use it with that method...

Thanks :)

**Reply by eliasdabbas** [reply]
Great to know you're having fun with it!
At the moment there is no way to use proxies, but you can try the following:

1. Change request headers: 
```python
import advertools as adv
adv.robotstxt.headers  # <-- change this dictionary as needed
```
2. Manually download the robots file (right click, save), and then run `robotstxt_to_df ` on the absolute file path on your local computer
3. Try running from a different location using Google Colab, Kaggle, or any other online notebook service.
Hope this helps!
